### PR TITLE
feat: add react-calendar package and fix types for emotion theme

### PR DESCRIPTION
**Pull request by SimeonC**
Completes #103

CleanShot 2023-02-09 at 13 23 43

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @tablecheck/tablekit-css@2.1.0-canary.159.4349577749.0
  npm install @tablecheck/tablekit-react-datepicker@2.1.0-canary.159.4349577749.0
  npm install @tablecheck/tablekit-react-fit-content-textarea@2.1.0-canary.159.4349577749.0
  npm install @tablecheck/tablekit-react-select@2.1.0-canary.159.4349577749.0
  npm install @tablecheck/tablekit-react@2.1.0-canary.159.4349577749.0
  # or 
  yarn add @tablecheck/tablekit-css@2.1.0-canary.159.4349577749.0
  yarn add @tablecheck/tablekit-react-datepicker@2.1.0-canary.159.4349577749.0
  yarn add @tablecheck/tablekit-react-fit-content-textarea@2.1.0-canary.159.4349577749.0
  yarn add @tablecheck/tablekit-react-select@2.1.0-canary.159.4349577749.0
  yarn add @tablecheck/tablekit-react@2.1.0-canary.159.4349577749.0
  ```
</details>

### CODE REVIEW

File: system/react-datepicker/src/Weeks/Week.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import styled from '@emotion/styled';
+import { action } from '@storybook/addon-actions';
+import { Story, Meta } from '@storybook/react';
+
+import { useDatePickerContext, Root } from '../Root';
+
+import { Week } from './index';
+
+export default {
+  title: 'TableKit/DatePicker/Week',
```
Comment (daniellizik) [inline system/react-datepicker/src/Weeks/Week.stories.tsx:10]: Screenshot 2023-02-17 at 13 49 43

not sure where this is in the diff (or if its even this PR) but in storybook there's two datepicker stories

File: system/react-datepicker/src/Root.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    maxDate: undefined | Date;
+  };
+
+const RootContext = React.createContext<null | ContextProps>(null);
+
+export function useDatePickerContext(): ContextProps {
+  const context = React.useContext(RootContext);
+  if (!context)
+    throw new Error(
+      '@tablecheck/tablekit-datepicker hook or component was used outside of Root element'
```
Comment (daniellizik) [inline system/react-datepicker/src/Root.tsx:31]: should this match the package name?
```
"name": "@tablecheck/tablekit-react-datepicker",
```

File: system/react-datepicker/src/Root.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      context
+        .getForwardProps({
+          calendars: context.calendars,
+          offset: (newYear - currentYear) * 12
+        })
+        .onClick({ defaultPrevented: false });
+  };
+}
+
+function composeSafeDate(date: Date | undefined) {
```
Comment (daniellizik) [inline system/react-datepicker/src/Root.tsx:78]: its called safe because the input can be undefined?

File: system/react-datepicker/src/Root.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  if (!date) return date;
+  const safeDate = new Date(date.getTime());
+  safeDate.setHours(0, 0, 0, 0);
+  return safeDate;
+}
+
+type InteractionDate = undefined | Date;
+function interactionDateReducer(
```
Comment (daniellizik) [inline system/react-datepicker/src/Root.tsx:86]: the fn above this is a "composer", this one is a "reducer" but they appear to be performing very similar logic. intended?

Comment (daniellizik) [reply system/react-datepicker/src/Root.tsx:86]: oh nvm i see how it's used

File: system/react-datepicker/src/Weeks/Week.stories.tsx (hunk 1 above)
Comment (SimeonC) [reply system/react-datepicker/src/Weeks/Week.stories.tsx:10]: Yea, it's a bit weird, one is a "story" and the other is the folder :shrug

File: system/react-datepicker/src/Root.tsx (hunk 4 above)
Comment (SimeonC) [reply system/react-datepicker/src/Root.tsx:86]: Nice catch, this one should really use the compose one.

File: system/react-datepicker/src/Root.tsx (hunk 3 above)
Comment (SimeonC) [reply system/react-datepicker/src/Root.tsx:78]: Good point, I'll change it to just me `composeDate` as the `safe` is a bit unneccesary. I think I had a habit at one point of using `safe` for things like that.

Comment (daniellizik) [reply system/react-datepicker/src/Root.tsx:78]: oh, well, i think the naming is fine. i wasn't being rhetorical, just wanted to clarify 😅

Comment (SimeonC) [reply system/react-datepicker/src/Root.tsx]: It's kinda an old habit of mine so it was good to point out. `safe` in the name is kinda ambiguous if you aren't me, so it's better to just remove it. The types tell us better what it's doing. Have Updated.

File: system/react-datepicker/src/Root.tsx (hunk 2 above)
Comment (SimeonC) [reply system/react-datepicker/src/Root.tsx]: Thanks, Have updated

File: system/react-datepicker/src/Weeks/Day.tsx [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    'aria-selected': isSelected ? 'true' : undefined,
+    children: (
+      <span role="presentation" aria-hidden="true">
+        {date.getDate()}
+      </span>
+    )
+  };
+}
+
+export function Day({ dateObj }: { dateObj: DateObj }): JSX.Element {
```
Comment (daniellizik) [inline system/react-datepicker/src/Weeks/Day.tsx:271]: for the booking form we need to be able to assign different "states" to individual days. https://tablecheck.atlassian.net/browse/BEXS-1030

a venue can be closed on a day for various reasons: holiday, sheet settings, closing the entire venue, etc. is it possible in the current implementation to do these?

1. pass in totally custom `data-*` attribs to the Day component for testing in downstream apps, by hooking into `useDayProps` or some other means
2. use the `<Day />` component to compose bespoke child elements in `<Week />`? for instance, putting icons in the button or providing new styles

cc @tirsop

Comment (SimeonC) [reply system/react-datepicker/src/Weeks/Day.tsx:271]: Would `isDayDisabled={(date) => false}` on the month/root be enough for that use case? I think that should be easy enough to do 🤔 

As for the current implementation you'd have to basically rebuild the `Month` component from scratch.

Comment (SimeonC) [reply system/react-datepicker/src/Weeks/Day.tsx:271]: @daniellizik (cc @tirsop) Re-reading the updated comment I have these thoughts;

> 1. pass in totally custom data-* attribs to the Day component for testing in downstream apps, by hooking into useDayProps or some other means

I kinda get what you want to do that for, but seeing as we effectively only have 2 styles in TK - normal and disabled I think the example test should look more like the following code snippet. This also aligns closer with cypress best practices.

```ts
cy.get(`[data-date=${noSheetDate}]`).should('be.disabled')
```

> 2. use the <Day /> component to compose bespoke child elements in <Week />. for instance, putting icons in the button or providing new styles

AFAIK we do not want to support this in TableKit (specifically for date-picker and other more complicated components, for primitives it's OK), if there's styling changes they need to run through the full TK process not be overwritten in project level 🤔

**Review by daniellizik — APPROVED** [review]

File: system/react-datepicker/src/Weeks/Day.tsx (hunk 5 above)
Comment (daniellizik) [reply system/react-datepicker/src/Weeks/Day.tsx]: ok, fine for now

**Review by daniellizik — APPROVED** [review]

**Review by daniellizik — APPROVED** [review]

**Review by SashaShostyr — APPROVED** [review]

**Review by daniellizik — APPROVED** [review]

**Review by SashaShostyr — APPROVED** [review]

**Pull request merged**